Fix: Problems with customer groups category discounts

Codencode opened the pull request:
The added method is the one that previously existed in the "AdminProducts" controller, which has been removed. I added the method to the "AdminGroupsControllerCore.php" controller because it is now only used there.

<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project!

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/#pull-requests

For type and category see:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#type--category
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 9.0.x
| Description?      | see: https://github.com/PrestaShop/PrestaShop/issues/37441
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | see: https://github.com/PrestaShop/PrestaShop/issues/37441
| UI Tests          | https://github.com/SiraDIOP/ga.tests.ui.pr/actions/runs/17669916911
| Fixed issue or discussion?     | Fixes #37441 #39477
| Related PRs       |  #38199
| Sponsor company   | @Codencode

Codencode commented:
Hello @boherm,
there is a problem similar to this one (#37991), the only difference is that it occurs on a different page.
You can check this comment to better understand: https://github.com/PrestaShop/PrestaShop/issues/37991#issuecomment-2643082707

What do you suggest?

boherm commented:
@Codencode, oh you're right. I prefer to block your PR for now, just to avoid early merging. Maybe we can fix that otherwise. 🤔 I will check that tomorrow ;)

boherm reviewed (dismissed):
I'm blocking this PR, maybe we can fix also the https://github.com/PrestaShop/PrestaShop/issues/37991 in one.

Codencode commented:
@boherm,
Okay, I asked for your opinion precisely to understand if we can address both issues.
Thanks for the feedback!

boherm commented:
Hi @Codencode, 
So... to me, I think that we need to add your `ajaxProcessGetCategoryTree` function in `PrestaShopBundle\Controller\Admin\Sell\Catalog\Product\ProductController` and configure a new route to use this function into `src/PrestaShopBundle/Resources/routing/sell/catalog/products/products.yml`. Then, edit the `tree.js` to use this new route 🤔
What do you think?

Codencode commented:
> src/PrestaShopBundle/Resources/routing/sell/catalog/products/products.yml

Hi @boherm,
I think it's a good idea; this way, we can always refer to this route.
In JS, can we retrieve the new route by passing parameters? I'm asking because, until now, I've never used routing in JS.

Codencode commented:
@boherm, okay, I'm running some tests, but when using `HelperTreeCategories `in the Product controller, which is a modern controller, I'm getting an error here:
https://github.com/PrestaShop/PrestaShop/blob/281be1ba0defea4367bfe7c4e825b3cf837f6715/classes/tree/Tree.php#L523-L535
because `$directory` is `null`.

This is due to the fact that here:
https://github.com/PrestaShop/PrestaShop/blob/281be1ba0defea4367bfe7c4e825b3cf837f6715/classes/tree/Tree.php#L321
`$this->getContext()->smarty->getTemplateDir(1)` is `null`.

I created a temporary branch with the code I produced that is causing the error. This is the commit:
https://github.com/Codencode/PrestaShop/commit/bb787cdfeecf0707ded4c2413f17fa6e8f5e44ac

Let me know if you have any ideas on how to fix this or if there's something I'm doing wrong.

Thanks!

boherm commented:
Yes, in modern part of PS, we need to use twig instead of smarty and also avoid the use of all older classes in legacy classes folder. And if I remember correctly, smarty isn't loaded in Symfony context anyway. 🤔
So, you need to adapt your code to fully use modern part and maybe rewrote the smarty template `subtree_associated_categories.tpl` into a twig one.
Also maybe another solution, I just saw that we have a `src/Core/Domain/Category/Query/GetCategoriesTree.php` query (that is handled by `src/Adapter/Category/QueryHandler/GetCategoriesTreeHandler.php` that is used in `CategoryController::getCategoriesTreeAction` but I have some doubts about the `AdminSecurity` attribute 🤔

Codencode commented:
I am trying to use Smarty with the current templates because this code is intended for Legacy controllers.
I am experimenting with a solution that is not ideal, as it requires modifying some files in the classes/tree path.
I am attempting to do this because it is, after all, a transitional solution to be used in the old Legacy controllers.
I imagine that when these controllers are migrated to Symfony, Twig will then be used instead.

Codencode commented:
@boherm 

I managed to make it work with Smarty, but as I mentioned, I had to modify the following files:
- classes/tree/Tree.php
- classes/tree/TreeToolbar.php
- classes/tree/TreeToolbarButton.php

It might not be the best solution.
Could you take a look at it? Just to get your opinion.

In the meantime, I’ll try to explore the second solution you proposed to see if I can implement it.
Thanks!

Codencode commented:
Indeed, the type `PrestaShopBundle\Form\Admin\Type\CategoryChoiceTreeType` already exists and is used for modifying categories. Therefore, in my opinion, converting `subtree_associated_categories.tpl` into a Twig file doesn’t make much sense, as we would end up with two Twig files doing the same thing. Correct me if I’m wrong.

In my view, for this solution, we should continue using `HelperTreeCategories`, which, however, requires the modifications I made to work with a modern (Symfony) controller.

Another solution would be to "restore" the Legacy controller `AdminCategoryController`, which would only include the method `ajaxProcessGetCategoryTree`.

What do you think?

Codencode commented:
@boherm, have you had a chance to review my solution?

Thank you!

Codencode commented:
@boherm, ok Thanks!

boherm commented:
> These are the 2 commits:
> 
> * [Codencode@bb787cd](https://github.com/Codencode/PrestaShop/commit/bb787cdfeecf0707ded4c2413f17fa6e8f5e44ac)
> * [Codencode@fb9745e](https://github.com/Codencode/PrestaShop/commit/fb9745ecfd7c3572734ad32c4188ad1e3002bd6d)

It's clear that it's not the best solution, but it could do the trick temporary to fix these issues.

Wdyt @jolelievre? 🤔

Codencode commented:
> > Indeed, the type `PrestaShopBundle\Form\Admin\Type\CategoryChoiceTreeType` already exists and is used for modifying categories. Therefore, in my opinion, converting `subtree_associated_categories.tpl` into a Twig file doesn’t make much sense, as we would end up with two Twig files doing the same thing. Correct me if I’m wrong.
> 
> Mmm, yeah maybe it's useless to rebuild this feature with twig. In modern controller, it's better to create a new component for form anyway.
> 
> > In my view, for this solution, we should continue using `HelperTreeCategories`, which, however, requires the modifications I made to work with a modern (Symfony) controller.
> 
> I agree with that.
> 
> > Another solution would be to "restore" the Legacy controller `AdminCategoryController`, which would only include the method `ajaxProcessGetCategoryTree`.
> 
> It's a simpler solution, it's true. But clearly, I will prefer to avoid this.

Alright, so let’s set aside the solution involving the use of a Legacy controller :)
If you have any suggestions regarding the code I proposed, I’m all ears!

Codencode commented:
I also thought of a solution very similar to this Codencode@fb9745e. It would suffice to perform the check with $this->getContext()->smarty->getTemplateDir(0) before $this->getContext()->smarty->getTemplateDir(1). This way, we would avoid checking the type of controller, i.e., $this->getContext()->controller instanceof \PrestaShop\PrestaShop\Core\Context\LegacyControllerContext.

I came up with this solution because I noticed that in this issue: https://github.com/PrestaShop/PrestaShop/issues/38133, there is a very similar problem, but it concerns the file type. However, the issue still stems from $this->getContext()->smarty->getTemplateDir(1) being executed before $this->getContext()->smarty->getTemplateDir(0), which causes an error.

I created this third solution, and you can find it here: https://github.com/PrestaShop/PrestaShop/commit/c5bc2d4d19ba7dfb6915cf34072189fd9098463b

I look forward to your feedback.

Thank you!

boherm commented:
Hi @Codencode!
Sorry for my ultra late response, I had totally forget this issue 😅

I'm not a big fan on what do you do in `Tree.php` file in your [TEST 03 commit](https://github.com/PrestaShop/PrestaShop/commit/c5bc2d4d19ba7dfb6915cf34072189fd9098463b). It looks like that you change the order of overrides in smarty templates 🤔 I don't really understand why you need this piece of code. Is it because of `HelperTreeCategories` used by your `categoryTreeAction` method?

Anyway, I'm sorry, I'm little confuse! 🫤

Codencode commented:
Hi @boherm,
No problem, don't worry.

After creating this PR, I opened another PR (https://github.com/PrestaShop/PrestaShop/pull/38199) related to a different issue, but I later realized it's also connected to this one.

From what I understand, the root issue is that Smarty only receives the "base" template, excluding any potential "overrides." You might get a better idea here: https://github.com/PrestaShop/PrestaShop/pull/38199/files

So, the solutions I proposed earlier aren't actually suitable because the problem lies upstream—I only realized this later while working on PR #38199.

Therefore, in my opinion, if PR #38199 gets approved, modifying the order of instructions might no longer be necessary.

Codencode committed:
Added new route 'admin_category_tree'

Added new route 'admin_category_tree' to handle the opening of the category tree

ps-jarvis commented:
<!-- PR_WORDING -->
This pull request seems to contain new translation strings. I have summarized them below to ease up review:

- `Admin.Orderscustomers.Feature`
    - [ ] `Successful update`

(Note: this is an automated message, but answering it will reach a real human)

Codencode commented:
@boherm
I performed the rebase, pulling in the merged changes, and updated the code since PR #38199 (which addresses part of ISSUE #37441) was merged.
The change introduces a new route to fetch the category tree, just as you had suggested.

Awaiting your feedback. Thanks for the guidance!

@kpodemski FYI

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php:
```diff
@@ -30,6 +30,8 @@
 
 use Doctrine\ORM\EntityManagerInterface;
 use Exception;
+use HelperTreeCategories;
```
Hi @Codencode 

PHPStan blocks the use of legacy classes in recent code by default, it's a rule to encourage us not to rely on legacy too much and develop new components in favor of old ones. Vut it's not always applicable so we pick the exceptions one by one and this is one of them.

So to allow this modification we need to update the baselines that lists all these exceptions, you can do so by running this command `php -d memory_limit=-1 ./vendor/bin/phpstan analyse -c phpstan.neon.dist --generate-baseline`

Codencode replied:
Hi @jolelievre,

While running the PHPStan analysis, the system wasn't adding the exceptions correctly for me. I probably have something not properly configured in my local environment, so I went ahead and added the exceptions manually in the file, and it worked.

Thanks a lot for the tip!!!

Codencode committed:
chore(phpstan): add baseline exception for legacy HelperTreeCategories usage in ProductController

FabienPapet commented:
It is easy in this case to not rely on the legacy too much. You can create an interface and an adapter class. Then use this interface in all symfony environment and also use DI.

Personal opinion : I also don't like using legacy constants inside controllers. You can also use DI for this.

Codencode commented:
Hello @FabienPapet,

Your idea would definitely be a great solution, but I'm not sure how to proceed. The issue is that `HelperTreeCategories` has various methods, and I wouldn't know how to handle it.

Is it worth doing this?
I don’t have much experience in this area, so I might be mistaken.

Thanks a lot!

FabienPapet commented:
After a closer look into the HelperCategoryTree, it looks like it's more complicated than expected :D Not sure it's worth.

If you still want an example of what I meant, you can look inside the Access Adapter (https://github.com/PrestaShop/PrestaShop/blob/develop/src/Adapter/Security/Access.php).

With Symfony, we don't use this adapter, but only the interfaces the service is implementing, so that *one day* we can replace the adapter class by another new service without changing anything. More, we don't add legacy usages, so we don't add technical debt here.

When applicable: there's an easy way to think about it: 
- One legacy class = one adapter inside symfony bundle.
- Create well-designed interfaces for all legacy methods.
- Implements inside adapters the new interfaces to the legacy and inject only the interfaces in the Symfony services so there's no legacy here.

Hope this helps :D

Codencode commented:
ping @jolelievre @boherm @FabienPapet @kpodemski

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php:
```diff
@@ -1877,4 +1879,50 @@ private function getShopIdFromShopContext(): ?int
+     * @return Response
+     */
+    #[AdminSecurity("is_granted('create', request.get('_legacy_controller')) || is_granted('update', request.get('_legacy_controller')) || is_granted('read', request.get('_legacy_controller'))")]
+    public function categoryTreeAction(
```
```suggestion
    public function legacyCategoryTreeAction(
```
Can you please rename this action, so we understand right away that it's related to legacy pages and ideally add a small comment explaining that it should be removed someday when all pages relying on this component will have been migrated

jolelievre commented on src/PrestaShopBundle/Resources/config/routing/admin/sell/catalog/products/product.yml:
```diff
@@ -520,3 +520,11 @@ admin_products_quantity:
+  defaults:
+    _controller: PrestaShopBundle\Controller\Admin\Sell\Catalog\Product\ProductController::categoryTreeAction
+    _legacy_controller: AdminProducts
+    _legacy_link: AdminProducts
```
```suggestion
    _legacy_link: AdminProducts:getCategoryTree
```
It's important to indicate the related action in the legacy link, it prevents this action from being confused with the default index action (when no action is defined it is assumed it is the index)

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Component/LegacyLayout/head_tag.html.twig:
```diff
@@ -64,6 +64,7 @@
   var admin_notification_push_link = adminNotificationPushLink = '{{ path('admin_common_notifications_ack') }}';
   var update_success_msg = '{{ 'Successful update'|trans({}, 'Admin.Notifications.Success') }}';
   var search_product_msg = '{{ 'Search for a product'|trans({}, 'Admin.Orderscustomers.Feature') }}';
+  var admin_category_tree = '{{ path('admin_category_tree') }}';
```
I'm not a fan of having this variable in the generic template, especially for a legacy url used only on very few pages Is there no other way to provide the correct address in the JS code? I have two different approches in mind:
- if you correctly setup the `_legacy_link` in the SF route then the call to the former legacy url should be automatically redirected so you won't even need to change something in the tree.js file
- you can use our `Router` js component to generate the URL, to do that you'll need to add the expose option like here https://github.com/PrestaShop/PrestaShop/blob/a248898655e56cbcc6c308a5f1c8752231624bae/src/PrestaShopBundle/Resources/config/routing/admin/_common.yml#L8 and rebuild the FOS routing file

jolelievre replied:
In case you go for the second solution the FOS file is generated with this command:

```js
php bin/console fos:js-routing:dump --format=json --target=admin-dev/themes/new-theme/js/fos_js_routes.json
```
And the JS router has never been used in the default theme yet, but it should be accessible if you use the JS components:
```js
window.prestashop.component.initComponents(
    [
      'Router',
    ],
  );
  window.prestashop.instance.router.generate(...);
```

Codencode replied:
@jolelievre
Fantastic! Indeed, with the first solution, there's no need to modify the tree.js file (as you rightly pointed out)!

Thank you so much! I'm going with this solution.
I'll publish the changes shortly

Codencode committed:
Refactor category tree route to legacyCategoryTreeAction

jolelievre reviewed (dismissed):
Thanks @Codencode 

So thanks to the legacy link it works as expected in the JS part? Did you test it?

If it works, then it's good for me.

Codencode committed:
Add @todo comment to legacyCategoryTreeAction for future removal

jolelievre reviewed (approved):
Thanks @Codencode

SiraDIOP reviewed (commented):
Hello @Codencode , 

Thanks for your PR, I tested it and it's good for me: 

https://github.com/user-attachments/assets/6e0bf94e-d435-4cee-8a05-dcb32fc9da84

https://github.com/user-attachments/assets/f43992d9-117a-4bad-9f43-d65f770dc5f9
